fix cogify

emileten opened the pull request:
The cogify code had deprecated CMR logic in it and only supports an HDF5 -> COG case with pretty custom conversion settings. 

For the purpose of https://github.com/MAAP-Project/ZenHub/issues/709 : 

- [x] removing CMR logic from the cogify lambda
- [x] adding a geotiff -> cloud optimized geotiff case using the default rio cogeo settings. 

The lambda was tested locally, with SMCE credentials. 

Input event : 


```
    sample_event = {
        "collection": "ESACCI_Biomass_L4_AGB_V4_100m_2020",
        "remote_fileurl": "s3://maap-ops-workspace/nehajo88/Data/CCI_2020/N00E000_ESACCI-BIOMASS-L4-AGB-MERGED-100m-2020-fv4.0.tif",
        "upload": True,
    }
```

Which saved the COG to `s3://nasa-maap-data-store/file-staging/nasa-map/ESACCI_Biomass_L4_AGB_V4_100m_2020/N00E000_ESACCI-BIOMASS-L4-AGB-MERGED-100m-2020-fv4.0.tif`. I checked it's a valid COG with `rio cogeo validate ...`. 

One can also check a view at https://titiler-stac.maap-project.org/cog/viewer?url=s3://nasa-maap-data-store/file-staging/nasa-map/ESACCI_Biomass_L4_AGB_V4_100m_2020/N00E000_ESACCI-BIOMASS-L4-AGB-MERGED-100m-2020-fv4.0.tif.

emileten committed:
remove CMR logic from cogify and add a geotiff to COG event on top of the existing HDF5 to COG event

omshinde reviewed (approved):
The visualization looks good and validates the updates. Thanks

jjfrench committed:
chore: ESACCI Biomass changes

jjfrench committed:
chore: Missing add to config, oops

jjfrench commented:
I added the inventory files under my public prefix accessible in the ADE

create the CSVs using:
```
(echo "s3_path"; aws s3 ls s3://maap-ops-workspace/nehajo88/Data/CCI_2020/ --recursive | grep ".tif$" | awk '{print "s3://maap-ops-workspace/" $4}') > CCI_2020.csv
aws s3 cp CCI_2020.csv s3://maap-ops-workspace/jjfrench/CCI_2020.csv
```

I can update the stepfn jsons later tonight

emileten commented:
Nice thanks @jjfrench. If you can't get to it by then I will later my day.

emileten committed:
modify stpfn input jsons for inventory

emileten committed:
flag versions in title, not using user shared bucket

emileten committed:
add csv url key

emileten commented:
@jjfrench here is the status of this. 


- I am not working on the `test` database anymore because I accidentally broke the associated secret, so I am working on `dev` and deployed this branch to the `dev` pipeline. 
- I inserted `ESACCI_Biomass_L4_AGB_V4_100m_2017`, just the collection. 
- Item insertion passes the inventory, and fails at `cogify`. It looks that we don't have the right permission set up in SMCE. [Here is the error](https://us-west-2.console.aws.amazon.com/states/home?region=us-west-2#/v2/executions/details/arn:aws:states:us-west-2:916098889494:execution:maap-data-pipelines-dev-stepfunction-cogify:ESACCI_Biomass_L4_AGB_V4_100m_2017-d021b3e8-f4a4-4d59-8fb0-97dd708404a9) (pasting it below as well)

```
{
  "errorMessage": "An error occurred (403) when calling the HeadObject operation: Forbidden",
  "errorType": "ClientError",
  "requestId": "eb54ed96-e524-436f-990b-33cda4cd9d79",
  "stackTrace": [
    "  File \"/app/handler.py\", line 217, in handler\n    downloaded_filename = download_file(file_uri=filename)\n",
    "  File \"/app/handler.py\", line 77, in download_file\n    s3.download_file(bucket, path, filename)\n",
    "  File \"/usr/local/lib/python3.9/site-packages/boto3/s3/inject.py\", line 190, in download_file\n    return transfer.download_file(\n",
    "  File \"/usr/local/lib/python3.9/site-packages/boto3/s3/transfer.py\", line 326, in download_file\n    future.result()\n",
    "  File \"/usr/local/lib/python3.9/site-packages/s3transfer/futures.py\", line 103, in result\n    return self._coordinator.result()\n",
    "  File \"/usr/local/lib/python3.9/site-packages/s3transfer/futures.py\", line 266, in result\n    raise self._exception\n",
    "  File \"/usr/local/lib/python3.9/site-packages/s3transfer/tasks.py\", line 269, in _main\n    self._submit(transfer_future=transfer_future, **kwargs)\n",
    "  File \"/usr/local/lib/python3.9/site-packages/s3transfer/download.py\", line 354, in _submit\n    response = client.head_object(\n",
    "  File \"/usr/local/lib/python3.9/site-packages/botocore/client.py\", line 530, in _api_call\n    return self._make_api_call(operation_name, kwargs)\n",
    "  File \"/usr/local/lib/python3.9/site-packages/botocore/client.py\", line 964, in _make_api_call\n    raise error_class(parsed_response, operation_name)\n"
  ]
}
```

jjfrench commented:
@emileten Thanks for changing those JSONs, I'll take a look at the permissions and get updated bucket policies in today

jjfrench committed:
fix: permissions policy patch

jjfrench committed:
style: lint...

emileten commented:
@jjfrench ran successfully one iteration of the cogify step. 

I just tried running the full pipeline, which triggers all iterations of the cogify step. 

There are a couple failures and successes so far. Here is one example of error : 


```
{
  "errorMessage": "TIFFAppendToStrip:Seek error at scanline 1792",
  "errorType": "CPLE_AppDefinedError",
  "requestId": "ad4f7f14-f620-4032-b712-1e1e8f0c1eb6",
  "stackTrace": [
    "  File \"/app/handler.py\", line 230, in handler\n    output_locations = geotiff_to_cog(\n",
    "  File \"/app/handler.py\", line 196, in geotiff_to_cog\n    cog_translate(\n",
    "  File \"/usr/local/lib/python3.9/site-packages/rio_cogeo/cogeo.py\", line 401, in cog_translate\n    copy(tmp_dst, dst_path, copy_src_overviews=True, **dst_kwargs)\n",
    "  File \"/usr/local/lib/python3.9/site-packages/rasterio/env.py\", line 451, in wrapper\n    return f(*args, **kwds)\n",
    "  File \"rasterio/shutil.pyx\", line 139, in rasterio.shutil.copy\n",
    "  File \"rasterio/_err.pyx\", line 221, in rasterio._err.exc_wrap_pointer\n"
  ]
}
```

It's not always the same error. Here is another one 

```
{
  "errorMessage": "TIFFRewriteDirectory:Error fetching directory count",
  "errorType": "CPLE_AppDefinedError",
  "requestId": "494acf29-a157-463b-8b30-c7f0c5ab2c53",
  "stackTrace": [
    "  File \"/app/handler.py\", line 230, in handler\n    output_locations = geotiff_to_cog(\n",
    "  File \"/app/handler.py\", line 196, in geotiff_to_cog\n    cog_translate(\n",
    "  File \"/usr/local/lib/python3.9/site-packages/rio_cogeo/cogeo.py\", line 401, in cog_translate\n    copy(tmp_dst, dst_path, copy_src_overviews=True, **dst_kwargs)\n",
    "  File \"/usr/local/lib/python3.9/site-packages/rasterio/env.py\", line 451, in wrapper\n    return f(*args, **kwds)\n",
    "  File \"rasterio/shutil.pyx\", line 139, in rasterio.shutil.copy\n",
    "  File \"rasterio/_err.pyx\", line 221, in rasterio._err.exc_wrap_pointer\n"
  ]
}
```

I don't know if these are reproducible errors or not, and I am also not sure if we have retries or not.

emileten commented:
So I picked one of these errors https://us-west-2.console.aws.amazon.com/states/home?region=us-west-2#/v2/executions/details/arn:aws:states:us-west-2:916098889494:execution:maap-data-pipelines-dev-stepfunction-cogify:ESACCI_Biomass_L4_AGB_V4_100m_2017-8373aacd-dbef-4f3d-9d27-b0fffcdeaa1b. The associated input file is 

`s3://maap-ops-workspace/nehajo88/Data/CCI_2017/S20W060_ESACCI-BIOMASS-L4-AGB-MERGED-100m-2017-fv4.0.tif`

If I run locally 

`rio cogeo create s3://maap-ops-workspace/nehajo88/Data/CCI_2017/S20W060_ESACCI-BIOMASS-L4-AGB-MERGED-100m-2017-fv4.0.tif S20W060_ESACCI-BIOMASS-L4-AGB-MERGED-100m-2017-fv4.0_COG.tif`

It succeeds and then I validate the COG successfully. So this has nothing to do with the file.

Trying to do that closer to what's happening in the step function, I installed the environment required for the cogify lambda https://github.com/MAAP-Project/maap-data-pipelines/tree/main/lambdas/cogify, and ran the handler with this input dict 

```
    sample_event = {
    "collection": "ESACCI_Biomass_L4_AGB_V4_100m_2017",
    "remote_fileurl": "s3://maap-ops-workspace/nehajo88/Data/CCI_2017/S20W060_ESACCI-BIOMASS-L4-AGB-MERGED-100m-2017-fv4.0.tif",
    "upload": True,
    "user_shared": False,
    "properties": None,
    "assets": {
        "csv": "s3://maap-ops-workspace/nehajo88/Data/CCI_2017/S20W060_ESACCI-BIOMASS-L4-AGB-MERGED-100m-2017-fv4.0.tif"
    },
    "product_id": "S20W060_ESACCI-BIOMASS-L4-AGB-MERGED-100m-2017-fv4.0"
    }
```

And it successfully converted to COG and uploaded the COG to the final destination, at 

```
s3://nasa-maap-data-store/file-staging/nasa-map/ESACCI_Biomass_L4_AGB_V4_100m_2017/S20W060_ESACCI-BIOMASS-L4-AGB-MERGED-100m-2017-fv4.0.tif
```

emileten committed:
avoid downloading, process in memory, make use of intermediate compression

emileten committed:
lint

emileten commented:
I implemented a couple changes suggested by Vincent to fix this. I tested them locally. I haven't deployed to AWS yet. 

- Don't download the file
- Process it in memory directly from the source S3 URI
- Write from the in mem file to the target S3 key directly
- make use of intermediate compression. 

@omshinde or @jjfrench if anyone wants to pick this up !

jjfrench committed:
fix: environment variable name

emileten committed:
fix: bump pydantic version

emileten committed:
add asset name and role

emileten committed:
fix regression

emileten committed:
expand regex

emileten committed:
change example

emileten committed:
add collection in cogify transfer output path

emileten committed:
update step fn input files

emileten commented:
- updated the step fn json files with asset name, role and type
- fixed a bug by bumping the pydantic version
- fixed a regression introduced recently with the `pystac_assets` object. 
- extended the regex definition

emileten commented:
Deployed to `dev` and published to the catalog.

emileten commented:
@jjfrench could you take a quick look and tell me if everything looks good in the collections/PR ? 

Looks like in the inventory files there are a couple non tif files, with extensions like `FLT64_CCI_2020_StratoForm_BOSQUE`. I think those will just trigger errors on their own without further issues.

emileten commented:
@jjfrench would you mind removing the undesired paths from the 2020 inventory ? And then I will delete-republish. 

status of the publication

- item count is good for 2017 and 2018 👍 
- 10 items are missing from 2019, I see a failed cogify run for 2019 here, that's probably the problem. I suspect rerunning will fix it, looking at the error https://us-west-2.console.aws.amazon.com/states/home?region=us-west-2#/v2/executions/details/arn:aws:states:us-west-2:916098889494:execution:maap-data-pipelines-dev-stepfunction-cogify:ESACCI_Biomass_L4_AGB_V4_100m_2019-174fd1c5-e895-4208-9734-1b6900b8286e 
- I realized there are SD files, so there is a pair of file for each location. Right now we're saving them as separate assets. We did the same thing with the previous version. I remember we had the standard deviation for ICEsat2 -- I think those were separate items ? Not sure. We could have it as a future improvement to save them in the same item.

jjfrench commented:
Oh interesting, it was only a grep ".tifs" so not entirely sure what happened there. Also, not sure if cogify is passing along the "asset_" keys in the stepfn dict - none of the items.item_assets got named as intended but it worked out all the same. Everything in so far looks correct

[Edit] oh, I saw the comment on the ticket. One's under a different prefix. I can also add in the SD data in the csv if it makes sense now / I reran that failed 2019, should be good

jjfrench commented on data/step_function_inputs/ESACCI_Biomass_L4_AGB_V4_100m_2017.json:
```diff
@@ -0,0 +1,14 @@
+  "user_shared": false,
+  "cogify": true,
+  "csv_file_url_key": "s3_path",
+  "asset_roles": ["data"],
+  "asset_media_type": {
+          "tif": "image/tiff; application=geotiff; profile=cloud-optimized"
+    },
+  "asset_name": "tif"
```
I believe these aren't passed through cogify (not necessarily needed though)

emileten replied:
I think it's needed? We're stuck with one asset name otherwise. 

Just pushed a change that ensures cogify passes its inputs along to the next step.

emileten committed:
correct inventory file

emileten committed:
make sure cogify passes its inputs along

emileten committed:
format

emileten commented:
- cogify passes along its input to the next step (hopefully).
- fixed the 2020 inventory file, now in emileten workspace, not jjfrench. 
- format.

emileten committed:
fix handler, csv asset should be something optional

emileten committed:
fix input jsons

emileten commented:
Merging. Reminders : 

- not deployed to test yet (backend deployment was teared down)
- we probably want the standard deviation and estimates in the same items as different assets.

emileten commented:
@jjfrench I remember you were mentioning after this PR got deployed, cogify was very slow. 

I thought maybe it's due to the compression we allow for.

So I just tested locally the following 

```
time rio cogeo create s3://maap-ops-workspace/nehajo88/Data/CCI_2020/S50W070_ESACCI-BIOMASS-L4-AGB_SD-MERGED-100m-2020-fv4.0.tif test_create.tif
```
versus 

```
time rio cogeo create --in-memory --allow-intermediate-compression  s3://maap-ops-workspace/nehajo88/Data/CCI_2020/S50W070_ESACCI-BIOMASS-L4-AGB_SD-MERGED-100m-2020-fv4.0.tif test_create_v2.tif
```

(The difference is `--allow-intermediate-compression` and `--in memory`)

And the processing time is the same. I think it's just slow to COG these things and there isn't much we can do ?
